modules/core/arrow-test/src/main/kotlin/arrow/test/laws/MonadDeferLaws.kt

```diff
@@ -31,9 +36,132 @@ object MonadDeferLaws {
+      Law("Sync laws: stack safety over repeated attempts") { SC.stackSafetyOverRepeatedAttempts(5000, EQ) },
+      Law("Sync laws: stack safety over repeated maps") { SC.stackSafetyOnRepeatedMaps(5000, EQ) }
     )
 
+  fun <F> MonadDefer<F>.delayConstantEqualsPure(EQ: Eq<Kind<F, Int>>): Unit {
+    forAll(genIntSmall()) { x ->
+      delay { x }.equalUnderTheLaw(just(x), EQ)
+    }
+  }
+
+  fun <F> MonadDefer<F>.deferConstantEqualsPure(EQ: Eq<Kind<F, Int>>): Unit {
+    forAll(genIntSmall()) { x ->
+      defer { just(x) }.equalUnderTheLaw(just(x), EQ)
+    }
+  }
+
+  fun <F> MonadDefer<F>.deferUnsafeConstantRightEqualsPure(EQ: Eq<Kind<F, Int>>): Unit {
+    forAll(genIntSmall()) { x ->
+      deferUnsafe { x.right() }.equalUnderTheLaw(just(x), EQ)
+    }
+  }
+
+  fun <F> MonadDefer<F>.deferUnsafeConstantLeftEqualsRaiseError(EQERR: Eq<Kind<F, Int>>): Unit {
+    forFew(5, genThrowable()) { t ->
+      deferUnsafe { t.left() }.equalUnderTheLaw(raiseError(t), EQERR)
+    }
+  }
+
+  fun <F> MonadDefer<F>.delayThrowEqualsRaiseError(EQERR: Eq<Kind<F, Int>>): Unit {
+    forFew(5, genThrowable()) { t ->
+      delay { throw t }.equalUnderTheLaw(raiseError(t), EQERR)
+    }
+  }
+
+  fun <F> MonadDefer<F>.propagateErrorsThroughBind(EQERR: Eq<Kind<F, Int>>): Unit {
+    forFew(5, genThrowable()) { t ->
+      delay { throw t }.flatMap<Int, Int> { a: Int -> just(a) }.equalUnderTheLaw(raiseError(t), EQERR)
+    }
+  }
+
+  fun <F> MonadDefer<F>.deferSuspendsEvaluation(EQ: Eq<Kind<F, Int>>): Unit {
+    val sideEffect = SideEffect(counter = 0)
+    val df = defer { sideEffect.increment(); just(sideEffect.counter) }
+
+    Thread.sleep(10)
+
+    sideEffect.counter shouldBe 0
+    df.equalUnderTheLaw(just(1), EQ) shouldBe true
+  }
+
+  fun <F> MonadDefer<F>.delaySuspendsEvaluation(EQ: Eq<Kind<F, Int>>): Unit {
+    val sideEffect = SideEffect(counter = 0)
+    val df = delay { sideEffect.increment(); sideEffect.counter }
+
+    Thread.sleep(10)
+
+    sideEffect.counter shouldBe 0
+    df.equalUnderTheLaw(just(1), EQ) shouldBe true
+  }
+
+  fun <F> MonadDefer<F>.flatMapSuspendsEvaluation(EQ: Eq<Kind<F, Int>>): Unit {
+    val sideEffect = SideEffect(counter = 0)
+    val df = just(0).flatMap { sideEffect.increment(); just(sideEffect.counter) }
+
+    Thread.sleep(10)
+
+    sideEffect.counter shouldBe 0
+    df.equalUnderTheLaw(just(1), EQ) shouldBe true
+  }
+
+  fun <F> MonadDefer<F>.mapSuspendsEvaluation(EQ: Eq<Kind<F, Int>>): Unit {
+    val sideEffect = SideEffect(counter = 0)
+    val df = just(0).map { sideEffect.increment(); sideEffect.counter }
+
+    Thread.sleep(10)
+
+    sideEffect.counter shouldBe 0
+    df.equalUnderTheLaw(just(1), EQ) shouldBe true
+  }
+
+  fun <F> MonadDefer<F>.repeatedSyncEvaluationNotMemoized(EQ: Eq<Kind<F, Int>>): Unit {
+    val sideEffect = SideEffect()
+    val df = delay { sideEffect.increment(); sideEffect.counter }
+
+    df.flatMap { df }.flatMap { df }.equalUnderTheLaw(just(3), EQ)
```

Comment:
shouldBe true, else this is not checked

Reply:
Can't believe I forgot that :sweat_smile:. Added it and it works for io, reactor and rx2 but fails for DeferredK.

Reply:
That is a good question and might be an issue or feature of coroutines. Will test this more thoroughly later on.
A possible scenario would be that coroutines simply after finishing return their result instead of rerunning when awaited again

Reply:
Well it makes sense now... Thats just how coroutines work:
```kotlin
  var sideEffect = 0
  val a = GlobalScope.async(start = CoroutineStart.LAZY) {
    ++sideEffect
  }

  println("Started a: ${a.start()}")

  runBlocking {
    println("Result a: ${a.await()}")
  }

  val b = GlobalScope.async(start = CoroutineStart.LAZY) {
    a.await()
  }

  println("Started b: ${b.start()}")

  runBlocking {
    println("Result b: ${b.await()}")
  }

  println(sideEffect)
```

Which prints: 
```
Started a: true
Result a: 1
Started b: true
Result b: 1
1
```

So what should be done about this? This is not like the stack-safety issue, if someone were to try and build a retry/repeat scheduler that won't work with DeferredK. Also other code which reuses MonadDefer instances (and then gets a deferredK) simply won't work. And the more annoying problem here is that it would be very hard to track down to Coroutines in that case :/

I don't really know how to solve this, wrapping deferred in DeferredK into a function has other challenges (stack-safety etc) that might not work... Maybe someone at jb knows what to do here...

Reply:
Right, I forgot they are memoized -.- This cannot be solved by JBs as it'd require moving the stack labels, which is the same issue we have with monad comprehensions, and it's just a design choice instead.

cc @raulraja for ideas

Reply:
Ok so how about I create a ticket for this. Comment the test out for now with a fixme linked to the ticket, add a copy to io's tests (also with a comment and link) and let that ticket deal with it then.